README.md
# TradingView to OANDA

## Summary
I'm setting up a [web.py](https://webpy.org/) server that can run on PythonAnywhere (uploaded there through GitHub) which listens to TradingView alerts on a URL (i.e. webhook) and posts orders (and sets stop-losses, etc.) on OANDA through their API

## Things to consider
* A sell order isn't actually handled as a sell order, but as closing all open positions for the given `instrument`.
  * Note that if this signal is sent to OANDA when the markets are closed, it will be cancelled.
* A buy order is handled as a limit order with a Good-Til-Date 15 minutes in the future.
  * That means that if it is sent when the markets are closed, it will most likely cancel (unless of course the markets open within those 15 minutes)

## Supported parameters
### Required parameters
* `action` either `buy` or `sell`
* `ticker` a six-letter ticker without any separator (e.g. `EURUSD`)
* `close` the price you want to buy the instrument at (ignored for sells, see [Things to consider](#things-to-consider))
### Optional parameters and their defaults
* `units` the size of the position to open as a positive integer, in the currency of the instrument (e.g. in Euros for "EURUSD" or gold for "XAUEUR"). Defaults to `500`
* `trailing_stop_loss_percent` the percentage points below `close` on which to start the trailing stop loss as a positive decimal. (E.g. to set it 5% below, enter `0.05`.) Defaults to `0.01`, i.e. 1%. See [OANDA's documentation on trailing stops](https://www1.oanda.com/forex-trading/learn/capital-management/stop-loss) for more information.
* `take_profit_percent` the percentage points above `close` to set the take profit at as a positive decimal. When the market reaches this point, the position is automatically closed. Defaults to `0.06`, i.e. 6%
* `trading_type` either `live` or `practice`, used to select the respective API key and account ID from your `credentials.json` file. Defaults to `practice`
Any other parameters will not be handled, but are sent along. That means you will see them in your server logs.

## To-do
1. ~~Set up development environment~~
1. ~~Install and set up the web.py server~~
1. ~~Have it listen to POST requests (test with Postman)~~
1. ~~Parse the parameters sent in the request to variables that can be sent along to OANDA~~
1. ~~Authenticate with the OANDA API~~
1. ~~Send a test order to OANDA's paper trader~~
1. ~~Add stop-losses and other risk management techniques to the OANDA call~~
1. ~~Connect the local web.py server with the OANDA API call~~
1. ~~Test if a request from Postman to the web.py server running locally reaches OANDA correctly~~
1. ~~Find out why my limit order didn't go through in the beginning, but cancelled because it reaches the GTD. Maybe the market was closed? If that was the case the market won't move and there won't be any signals, right?~~
1. ~~Add some kind of verification so that only I can send orders through my webhook~~
1. ~~Move the server to PythonAnywhere so that TradingView can reach it~~
1. ~~Setup TradingView to send an alert to the server on PythonAnywhere~~
1. ~~Have the server return the result of a POST request for easier Postman debugging~~
1. ~~Have the server translate what TradingView sends as `{{ticker}}` and `{{close}}` to what OANDA expects as `instrument` and `price`~~
1. ~~Make `size` optional with a default value set server-side~~
1. ~~Add different processing for buy and sell alerts~~
1. ~~Fix the fact that `size` works for XAUEUR, but not for `EURUSD`~~
1. ~~Rename `close` to `price` in the alerts and everywhere else, in case someone wants to use a different price~~
1. ~~Add more alerts to TradingView once I have ironed out the JSON in the alerts~~
1. ~~Automatically set price precision decimals and remove it as a parameter that can be messed with~~
1. ~~Wait and see if one of those alerts from TradingView reaches OANDA correctly~~
1. ~~Add a mechanism that notifies me of every order (email, maybe?)~~
1. `git pull` the latest version to PythonAnywhere so that the email subject of sell orders is correct
1. Code cleanup, more object-oriented, maybe?
1. Think of a way to handle alerts based on the last candle of the day, which will most likely be triggering buys/sells when the markets are closed

## Prerequisites
* `pip install web.py`
* `pip install requests`
* `pip install sendgrid`, if you choose to use [SendGrid's mail API](https://sendgrid.com/docs/API_Reference/api_v3.html) to send order confirmations from PythonAnywhere to yourself
* a `credentials.json` file containing a JSON like this:

```json
{
  "oanda_practice": {
    "api_key": "<YOUR PRACTICE ACCOUNT API KEY>",
    "account_id": "<YOUR PRACTICE ACCOUNT ID>"
  },
  "oanda_live": {
    "api_key": "<YOUR LIVE ACCOUNT API KEY>",
    "account_id": "<YOUR LIVE ACCOUNT ID>"
  },
  "sendgrid": {
    "api_key": "<YOUR SENDGRID API KEY>",
    "email_address": "<THE ADDRESS TO SEND FROM AND TO>"
  }
}
```

## Authentication
The webhook is only available on `/webhook/<ACCESS_TOKEN>` where `<ACCESS_TOKEN>` is one of the items in the JSON list in `access_tokens.json`. Create that file with the following contents:

```json
[
  "<ACCESS_TOKEN>"
]
```

Make sure to replace `<ACCESS_TOKEN>` with an access token of your choice (e.g. [have DuckDuckGo generate one for you](https://duckduckgo.com/?q=password+64))

Note that these are the only valid endpoints. The server won't tell you anything if you're trying to reach it somewhere else.

## Set up PythonAnywhere
1. Create an account, if you don't have one already
1. `git clone` this repository through a Bash console. (I'll assume you clone into `/home/<USERNAME>/tradingview-to-oanda` for the rest of the setup.)
1. Go to the "Files" tab and upload your `credentials.json` and `access_tokens.json` to the repository's directory
1. Create a virtualenv (I named it "tradingview-to-oanda" and used Python 3.7, see the [docs](https://help.pythonanywhere.com/pages/Virtualenvs)):
  ```
  mkvirtualenv tradingview-to-oanda --python=/usr/bin/python3.7
  ```
1. Open the "Web" tab and create a custom web app
1. Set the following parameters:
  * __Source code__ = The directory you cloned the repository into
  * __Working directory__ = The same directory
  * __Virtualenv__ = The directory of the virtualenv you created, probably something like `/home/<USERNAME>/.virtualenvs/tradingview-to-oanda`
1. Then open the __WSGI configuration file__ linked from the Web tab
1. Comment the `HELLO WORLD` bit (⌘+/ works here)
1. Add this to the bottom:
  ```
  # +++++++++++ web.py +++++++++++
  # Added by me, based on https://help.pythonanywhere.com/pages/WebDotPyWSGIConfig
  import sys
  sys.path.append('/home/<USERNAME>/tradingview-to-oanda')
  from server import app
  application = app.wsgifunc()
  ```
1. Save the file and go back to the Web tab
1. Click the green Reload button
1. Send a valid POST request to https://<USERNAME>.pythonanywhere.com/webhook/<ACCESS_TOKEN> and see if it reaches OANDA

## Add alerts to TradingView
1. Add an alert to whatever condition you want, for example [a custom-built `study`](https://www.tradingview.com/pine-script-docs/en/v4/annotations/Alert_conditions.html)
1. Check the __Webhook URL__ box and paste the URL of your PythonAnywhere webhook including access token
1. Make sure that the __Message__ is a valid JSON containing at least:
  ```json
  {"action":"buy","ticker":"{{ticker}}","close":{{close}}}
  ```
  Note:
  * The server is built in such a way that what TradingView sends as `{{ticker}}` is translated to what OANDA expects as `instrument` and `close` is translated to `price`
  * If you want to sell every position you have for this ticker, use "`sell`" as `action`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
*

## [0.0.1] - 2020-04-02
* Added Summary and To-do to README


oanda.py
import datetime
import json
import logging
import os.path

import requests

def get_datetime_offset(offset_minutes=15):
    # In RFC3339
    now = datetime.datetime.utcnow()
    now_plus_offset = now + datetime.timedelta(minutes=offset_minutes)
    return "{}Z".format(now_plus_offset.isoformat("T"))

def get_datetime_now():
    now = datetime.datetime.utcnow()
    return "{}Z".format(now.isoformat("T"))

def get_credentials(trading_type):
    loc = "oanda.py:get_credentials"

    try:
        with open("credentials.json") as credentials_json:
            credentials = json.load(credentials_json)["oanda_{}"
                .format(trading_type)]
    except Exception as e:
        logging.exception("{}: Could not read {} credentials from "
                          "credentials.json: ".format(loc, trading_type, e))
        raise

    return credentials

def get_base_url(trading_type):
    return "https://api-fx{}.oanda.com".format(
        "trade" if trading_type == "live" else "practice")

def get_accounts(trading_type="practice"):
    # https://developer.oanda.com/rest-live-v20/account-ep/
    loc = "oanda.py:get_accounts"

    try:
        credentials = get_credentials(trading_type)

        url = "{}/v3/accounts".format(get_base_url(trading_type))
        headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer {}".format(credentials["api_key"]),
        }

        response = requests.request("GET", url, headers=headers)

        return response
    except Exception as e:
        logging.exception("{}: Could not get {} accounts from the OANDA API: {}"
                          .format(loc, trading_type, e))
        raise

def get_instruments(trading_type="practice"):
    # https://developer.oanda.com/rest-live-v20/account-ep/
    loc = "oanda.py:get_instruments"

    try:
        credentials = get_credentials(trading_type)

        url = "{}/v3/accounts/{}/instruments".format(
            get_base_url(trading_type),
            credentials["account_id"])

        headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer {}".format(credentials["api_key"]),
        }

        instruments_response = requests.request("GET", url, headers=headers)
        instruments = json.loads(instruments_response.text.encode("utf8"))
        return instruments
    except Exception as e:
        logging.exception("{}: Could not get {} instruments from the OANDA "
                          "API: {}".format(loc, trading_type, e))
        raise

def get_price_precision(instrument, trading_type="practice"):
    price_precisions = get_price_precisions(trading_type)
    return price_precisions[instrument]

def get_price_precisions(trading_type="practice"):
    price_precisions_file = "price_precisions.json"

    if os.path.isfile(price_precisions_file):
        price_precisions = load_price_precisions(price_precisions_file)
        # Note that there is just 1 list for both trading types,
        # although I do not think they differ
    else:
        price_precisions = save_price_precisions(
            price_precisions_file, trading_type)

    return price_precisions

def save_price_precisions(price_precisions_file, trading_type="practice"):
    instruments = get_instruments(trading_type)

    price_precisions = {instrument["name"]:instrument["displayPrecision"] for
        instrument in instruments["instruments"]}

    with open("price_precisions.json", "w") as price_precisions_json:
        json.dump(price_precisions, price_precisions_json, indent=2,
                  sort_keys=True)

    return price_precisions

def load_price_precisions(price_precisions_file):
    with open(price_precisions_file) as price_precisions_json:
        price_precisions = (json.load(price_precisions_json))
    return price_precisions

def get_filtered_instruments(instrument_filter="EUR", trading_type="practice"):
    loc = "oanda.py:get_filtered_instruments"

    instruments = get_instruments(trading_type)
    filtered_instruments = list(filter(lambda i: instrument_filter in i,
                                   instruments))

    return filtered_instruments

def buy_order(instrument, units, price, trailing_stop_loss_percent,
              take_profit_percent, trading_type="practice",
              **kwargs):
    # https://developer.oanda.com/rest-live-v20/order-ep/
    loc = "oanda.py:buy_order"

    try:
        credentials = get_credentials(trading_type)
        price_decimals = get_price_precision(instrument, trading_type)

        url = "{}/v3/accounts/{}/orders".format(
            get_base_url(trading_type),
            credentials["account_id"]
        )

        # Convert the entered percentages to the absolute values OANDA expects
        trailing_stop_loss_distance = trailing_stop_loss_percent * price
        take_profit_price = price * (1 + take_profit_percent)

        payload = {
            "order": {
                "type": "LIMIT",
                "positionFill": "DEFAULT",
                "timeInForce": "GTD",
                "gtdTime": get_datetime_offset(15), # i.e. 15 m from now
                "instrument": instrument,
                "units": "{0:d}".format(units), # whole units
                "price": "{0:.{1}f}".format(price, price_decimals),
                "trailingStopLossOnFill": {
                    "distance": "{0:.{1}f}".format(trailing_stop_loss_distance,
                                                   price_decimals),
                    "timeInForce": "GTC",
                    "clientExtensions": {
                        "comment": "oanda.py/buy_order/trailing_stop_loss",
                        "tag": "trailing_stop_loss",
                        "id": "{}_trailing_stop_loss".format(get_datetime_now())
                    },
                },
                "takeProfitOnFill": {
                    "price": "{0:.{1}f}".format(take_profit_price, price_decimals),
                    "clientExtensions": {
                        "comment": "oanda.py/buy_order/take_profit",
                        "tag": "take_profit",
                        "id": "{}_take_profit".format(get_datetime_now())
                    },
                },
                "clientExtensions": {
                    "comment": "oanda.py/buy_order/entry",
                    "tag": "entry",
                    "id": "{}_entry".format(get_datetime_now())
                },
            }
        }

        payload_str = json.dumps(payload)
        headers = {
            "Content-Type": "application/json",
            "Authorization": "Bearer {}".format(credentials["api_key"]),
            "Accept-Datetime-Format": "RFC3339"
        }

        response = requests.request(
            "POST", url, headers=headers, data=payload_str)
    except Exception as e:
        logging.exception("{}: Could not send the buy order to OANDA: {}"
                          .format(loc, e))
        raise
    else:
        response_text = response.text.encode("utf8")
        response_json = json.loads(response_text)

        return response_json

def sell_order(instrument, trading_type, **kwargs):
    # https://developer.oanda.com/rest-live-v20/position-ep/
    loc = "oanda.py:sell_order"

    try:
        credentials = get_credentials(trading_type)

        url = "{}/v3/accounts/{}/positions/{}/close".format(
            get_base_url(trading_type),
            credentials["account_id"],
            instrument)

        payload = {
            "longUnits": "ALL",
            "longClientExtensions": {
                    "comment": "oanda.py/sell_order/close",
                    "tag": "close",
                    "id": "{}_close".format(get_datetime_now())
                },
            "shortUnits": "NONE",
        }

        payload_str = json.dumps(payload)
        headers = {
            "Content-Type":"application/json",
            "Authorization":"Bearer {}".format(credentials["api_key"]),
            "Accept-Datetime-Format":"RFC3339"}

        response = requests.request("PUT", url, headers=headers, data=payload_str)
    except Exception as e:
        logging.exception("{}: Could not send the sell order to OANDA: {}"
                          .format(loc, e))
        raise
    else:
        response_text = response.text.encode("utf8")
        response_json = json.loads(response_text)

        return response_json

if __name__ == "__main__":
    # Set logging parameters
    logging.basicConfig(level=logging.INFO)
    loc = "oanda.py"

    # Uncomment this bit to write all instruments and their price
    # precision—for the given trading type—to price_precisions.json, or
    # load them from that file if it exists
    # logging.info("{}: {}".format(loc, json.dumps(
    #     get_price_precisions(), indent=2, sort_keys=True)))

    # Uncomment this bit to send a buy order to OANDA
    order_response = buy_order(
        instrument="XAU_EUR",
        units=1, # i.e. 1 unit (bar?) of gold
        price=1486.891,
        trailing_stop_loss_percent=0.03, # as positive decimal
        take_profit_percent=0.06, # as positive decimal
        trading_type="practice"
    )

    logging.info("{}: {}".format(
        loc, json.dumps(order_response, indent=2, sort_keys=True)))


sendgrid_api.py
import json

import logging
from sendgrid import SendGridAPIClient
from sendgrid.helpers.mail import Mail

from oanda import get_datetime_now

def get_credentials():
    loc = "sendgrid_api.py:get_credentials"
    try:
        with open("credentials.json") as credentials_json:
            credentials = (json.load(credentials_json))["sendgrid"]
    except:
        logging.exception("{}: Could not read credentials from credentials.json"
                          .format(loc))
        raise

    return credentials

def send_mail(subject, message):
    # https://github.com/sendgrid/sendgrid-python

    loc = "sendgrid_api.py:send_mail"

    credentials = get_credentials()

    mail_to_send = Mail(
        from_email=credentials["email_address"],
        to_emails=credentials["email_address"],
        subject=subject,
        plain_text_content=message,
        html_content=None)

    try:
        sendgrid_client = SendGridAPIClient(credentials["api_key"])
        response = sendgrid_client.send(mail_to_send)

        return response
    except Exception as e:
        logging.exception("{}: Email could not be sent: {}".format(loc, e))
        raise

def success_mail(message):
    response = send_mail("TradingView to OANDA: Success", message)
    return response

def fail_mail(message):
    response = send_mail("TradingView to OANDA: Fail", message)
    return response

if __name__ == "__main__":
    # Set logging parameters
    logging.basicConfig(level=logging.INFO)
    loc = "sendgrid_api.py"

    response = send_mail(
        subject="SendGrid test from within {}".format(loc),
        message=get_datetime_now()
    )

    logging.info("{}: Email sent successfully. SendGrid answered with a {} "
                 "statuscode: {}".format(
        loc, response.status_code, response.body))

server.py
from copy import copy
import itertools
import json
from json.decoder import JSONDecodeError
import logging

import web

from oanda import buy_order, sell_order, get_datetime_now
from sendgrid_api import send_mail

def fill_defaults(post_data):
    loc = "server.py:fill_defaults"
    try:
        instrument = post_data["instrument"]
        price = float(post_data["price"])
    except KeyError as e:
        logging.exception("{}: One of the required parameters was missing: {}"
                          .format(loc, e))
        raise

    units = (
        int(post_data["units"])
        if "units" in post_data
        else 500) # e.g. in Euros for "EURUSD" or in gold for "XAUEUR"
    trailing_stop_loss_percent = (
        float(post_data["trailing_stop_loss_percent"])
        if "trailing_stop_loss_percent" in post_data
        else 0.01) # as positive decimal
    take_profit_percent = (
        float(post_data["take_profit_percent"])
        if "take_profit_percent" in post_data
        else 0.06) # as positive decimal
    trading_type = (
        post_data["trading_type"]
        if "trading_type" in post_data
        else "practice")

    return {
        "instrument": instrument,
        "units": units,
        "price": price,
        "trailing_stop_loss_percent": trailing_stop_loss_percent,
        "take_profit_percent": take_profit_percent,
        "trading_type": trading_type,
    }

def translate(post_data):
    loc = "server.py:translate"

    try:
        # Translate `ticker` to `instrument` by adding a "_"
        ticker = post_data.pop("ticker")
        if len(ticker) == 6:
            post_data["instrument"] = "{}_{}".format(ticker[:3], ticker[3:])
        else:
            raise ValueError("This ticker does not match my assumptions")

    except KeyError as e:
        logging.exception("{}: One of the parameters I am trying to translate "
                          "does not exist: {}".format(loc, e))
        raise
    except ValueError as e:
        logging.exception("{}: One of the parameters I am trying to translate "
                          "does not match my assumptions: {}".format(loc, e))
        raise

    return post_data

def post_data_to_oanda_parameters(post_data):
    loc = "server.py:post_data_to_oanda_parameters"

    try:
        translated_data = translate(post_data)
        filled_data = fill_defaults(translated_data)
    except Exception as e:
        logging.exception("{}: Could not translate data or fill all"
                          "defaults: {}".format(loc, e))
        raise

    return filled_data

class log:
    def __init__(self):
        self.content = ""

    def __str__(self):
        return str(self.content)

    def add(self, message):
        if len(self.content) > 0:
            self.content += "\n"

        self.content = "{}{}: {}".format(
            self.content, get_datetime_now(), message)

class webhook:
    def POST(self):
        loc = "server.py:webhook:POST"
        local_log = log()

        web_data = web.data()

        # Load the received JSON
        try:
            post_data = json.loads(web_data)
        except JSONDecodeError as e:
            error_message = ("{}: Request received with invalid JSON: {}".format(loc, e))
            logging.exception(error_message)

            local_log.add("ERROR:root:{}:\n{}".format(
                error_message, web_data))

            mail_response = send_mail(
                "TradingView to OANDA: Fail", str(local_log))
            local_log.add("INFO:ROOT:{}: I also sent you this log through "
                          "SendGrid. They replied with status code {}: {}"
                .format(loc, mail_response.status_code, mail_response.body))

            raise web.internalerror(local_log)

        info_message = "{}: Request received with valid JSON".format(loc)
        logging.info(info_message)
        local_log.add("INFO:ROOT:{}:\n{}".format(
                info_message, json.dumps(post_data, indent=2, sort_keys=True)))

        # And, if it could be loaded, translate it to OANDA parameters
        try:
            oanda_parameters = post_data_to_oanda_parameters(copy(post_data))
        except Exception as e:
            error_message = ("{}: Could not translate JSON to OANDA "
                             "parameters: {}".format(loc, e))
            logging.exception(error_message)

            local_log.add("ERROR:root:{}".format(error_message))

            mail_response = send_mail(
                "TradingView to OANDA: Fail", str(local_log))
            local_log.add("INFO:ROOT:{}: I also sent you this log through "
                          "SendGrid. They replied with status code {}: {}"
                .format(loc, mail_response.status_code, mail_response.body))

            raise web.internalerror(local_log)
        info_message = ("{}: Translated that to OANDA parameters".format(loc))
        logging.info(info_message)
        local_log.add("INFO:ROOT:{}:\n{}".format(
            info_message, json.dumps(
                oanda_parameters, indent=2, sort_keys=True)))

        # Then send those to OANDA as either a buy or sell order
        try:
            if post_data["action"] == "buy":
                order_response = buy_order(**oanda_parameters)
                mail_subject = ("TradingView to OANDA: Sent an order to buy {} "
                                "units of {}".format(
                    oanda_parameters["units"], oanda_parameters["instrument"]))
            elif post_data["action"] == "sell":
                order_response = sell_order(**oanda_parameters)
                mail_subject = ("TradingView to OANDA: Sent an order to close "
                                "all positions of {}".format(
                    oanda_parameters["instrument"]))
            else:
                raise ValueError("You did not specify whether you want to buy "
                                 "or sell")
        except Exception as e:
            error_message = ("{}: Could not send the order to OANDA because of "
                             "an error: {}".format(loc, e))
            logging.exception(error_message)

            local_log.add("ERROR:root:{}".format(error_message))

            mail_response = send_mail(
                "TradingView to OANDA: Fail", str(local_log))
            local_log.add("INFO:ROOT:{}: I also sent you this log through "
                          "SendGrid. They replied with status code {}: {}"
                .format(loc, mail_response.status_code, mail_response.body))

            raise web.internalerror(local_log)
        info_message = "{}: Sent order to OANDA".format(loc)
        logging.info(info_message)
        local_log.add("INFO:ROOT:{}:\n{}".format(info_message,
            json.dumps(order_response, indent=2, sort_keys=True)))

        # Mail the log and reply to the POST request
        mail_response = send_mail(mail_subject, str(local_log))
        local_log.add("INFO:ROOT:{}: I also sent you this log through "
                      "SendGrid. They replied with status code {}: {}".format(
            loc, mail_response.status_code, mail_response.body))

        web.header("Content-Type", "text/plain")
        return (local_log)

# Set logging parameters
logging.basicConfig(level=logging.INFO)
loc = "server.py"

# Load the list of access tokens and set webhook URLs for each one
try:
    with open("access_tokens.json") as access_tokens_json:
        access_tokens = (json.load(access_tokens_json))
except JSONDecodeError as e:
    logging.exception(
        "{}: Could not read tokens from access_tokens.json: {}".format(loc, e))
    raise

urls = tuple(itertools.chain.from_iterable(
    ["/webhook/{}".format(token)] + ["webhook"] for token in access_tokens))

# Set up the server
app = web.application(urls, globals())

if __name__ == "__main__":
    # Start the server
    logging.info("{}: Starting the server".format(loc))
    app.run()
